src/OTCNexus.sol

```diff
@@ -0,0 +1,283 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./OtcMath.sol";
+
+error Router__InvalidTokenAmount();
+error Router__InvalidPriceAmount();
+error Router__InvalidPriceMultiplier();
+error Router__InvalidDeadline();
+error Router__InvalidTokenAddresses();
+error Router__InvalidTokenAndAmountMismatch();
+error Router__AllowanceToken0TooLow();
+error Router__AllowanceToken1TooLow();
+error Router__ApproveToken0Failed();
+error Router__DepositToken0Failed();
+error Router__RfsRemoved();
+error Router__Amount1TooHigh();
+error Router__TransferToken0Failed();
+error Router__TransferToken1Failed();
+error Router__NotMaker();
+error Router__RefundToken0Failed();
+error Router__OnlyFixedOrDynamicAllowed();
+error Router__OnlyFixedPriceOrAmountAllowed();
+error Router__OnlyOneTokenAccepted();
+
+/**
+ * @title OTCNexus
+ * @notice core swap functionality
+ */
+contract OTCNexus is Ownable {
+    uint256 public rfsIdCounter;
+    mapping(uint256 => RFS) private idToRfs;
+    mapping(address => RFS[]) private makerRfs;
+
+    struct RFS {
+        uint256 id;
+        address maker;
+        address token0;
+        address[] tokensAccepted;
+        uint256 amount0;
+        uint256 amount1;
+        uint256 usdPrice; // Implies the fixed price per token0
+        uint256 priceMultiplier;
+        uint256 deadline;
+        bool isDynamic;
+        bool isDeposited;
+        bool removed;
+    }
+
+    event RfsUpdated(uint256 rfsId, uint256 amount0, uint256 amount1, uint256 deadline);
+    event RfsFilled(uint256 rfsId, address taker, uint256 amount0, uint256 amount1);
+    event RfsRemoved(uint256 rfsId);
+    event RfsCreated(
+        uint256 rfsId,
+        address maker,
+        address indexed token0,
+        address[] indexed tokensAccepted,
+        uint256 amount0,
+        uint256 usdPrice,
+        uint8 priceMultiplier,
+        uint256 deadline
+    );
+
+    /**
+     * @notice create an RFS where the price is either fixed to a USD price or it requests only a specific amount of tokens
+     * @param _token0 token offered
+     * @param _tokensAccepted token requested
+     * @param _amount0 amount of token offered
+     * @param _amount1 amount of token requested
+     * @param _usdPrice fixed price of the asset
+     * @param _deadline timestamp after which the RFS expires
+     * @param _isDeposited if the tokens are deposited or approved
+     * @return rfsId
+     */
+    function createFixedRfs(
+        address _token0,
+        address[] calldata _tokensAccepted,
+        uint256 _amount0,
+        uint256 _amount1,
```

Comment:
1. This is a current problem that we need to address. Since fixed should be allowed only for shitcoin that don't have oracle data. Maths implemented by @quantoor should solve this problem for various `tokensAccepted`.
2. Because we want to give users ability to accept various currencies instead of just one. This allows for more flexibility and more inclusions for various other `takers` as they don't have to previously swap to a specific currency.